Toggle token drawer

cpytel opened the pull request:
Closes #103. The moves the token drawer behind a toolbar button. 

![Drawer Token Open and Drag Out](https://user-images.githubusercontent.com/5015/90987283-c3e98d00-e557-11ea-8b66-ff3a1ec27e16.gif)

The drawer is automatically opened whenever you start to drag a token and closed when you stop.

![Drawer Token Start Drag](https://user-images.githubusercontent.com/5015/90987203-1aa29700-e557-11ea-975a-b9b7aa52b098.gif)

You can drag tokens back into the open drawer.

![Drawer Token Drag In](https://user-images.githubusercontent.com/5015/90987213-2b530d00-e557-11ea-817b-2e300f70a3c5.gif)

cpytel commented:
@jdbann this is "working" in general, but there is a z-index problem that I can't resolve. When dragging a token back onto the drawer, the token you're dragging appears underneath the drawer instead of on top of it. Dropping it works properly, and the drawer then closes as expected, but while you're holding onto the token it is under the drawer.

I think this is because the token container is being transformed, which according to https://coder-coder.com/z-index-isnt-working/ puts it on a different z-index plane. I tried moving the token drawer into the container in order to get it to be on the same container and plane, this didn't work as expected, it seemed to completely break dragging onto the drawer, it just ignored it. Secondarily, we also then need to transform the X and the Y to essentially make our own fixed positioning of the drawer and I couldnt' get the transform correct. I abandoned it because it broke the dropping on the drawer (dragging out still worked).

cpytel commented:
@jdbann ha! While working on this more I switched back to master, and it turns out this problem with mouseover also exists on master! Dragging a token over the drawer, it goes under the drawer and doesn't appear in the drawer until you drop it!

So, open to idea on how how to resolve, but it turns out this isn't a new problem only in this branch.

Next, I'm going to explore a solution of moving it into the dom of the drawer while you're dragging (but not actually stashing the token until you drop).

cpytel commented:
I've pushed up https://github.com/tableparty/table/pull/110/commits/413821710725d7ff09854ed1dc7c8bc0347aa680 which adds the token to the drawer dom once the drag enters the drawer.

I experimented with a few options, like keeping it absolutely positioned within the drawer until you drop it so it stays under your cursor, but I couldn't get it to work properly (I think because of the transforming that's happening in tokenContainer). 

The way it's working now isn't fantastic, but I actually do think it's an improvement over the existing drawer and the drawer dragging behavior in master. The code features some duplication and I haven't fully updated the specs yet, but I'd like to get your thoughts before continuing.

jdbann commented:
I haven't had a proper review yet, but from your comments here I was wondering if it's worth us going back to moving the tokens with mouse events rather than drag events again? It seems like we're having to bend drag and drop to our will a lot more than I was expecting.

I'm also open to getting rid of the drop-token-on-drawer behaviour if we're going for more of a token control approach as discussed in #104

cpytel commented:
I've fixed up all the specs and added animated gifs to the PR description, this is ready for review now.

I normally use Safari but while doing QA, I switched over to Chrome and discovered that the token zIndex issues from the map to the drawer aren't a problem in Chrome, so it looks pretty nice there. On Safari, it works better than the current drawer does on Safari.

cpytel commented:
@jdbann any issue with me merging this this weekend so it's in place for our next session?

jdbann commented on app/javascript/controllers/map/tokens_controller.js:
```diff
@@ -50,6 +50,9 @@ export default class extends Controller {
   startMoveToken(event) {
     event.stopPropagation()
+    if (this.hasDrawerTarget) {
+      this.drawerTarget.dispatchEvent(new CustomEvent('open'))
```
How do you feel about these events being defined elsewhere as constants? Or the event names?

Only asking as it seems to be the practice for similar use in Redux

cpytel replied:
Yeah, the thing that is interesting to me is that we have the ToggleController, so it feels like the ToggleController should actually have those constants or otherwise be more encapsulated, but I did some reading and emitting events seems to be the recommended way with stimulus.

Do you think they should live in constants.js file or something?

jdbann replied:
From reading through the rest of the PR, these events are used as the `data-action` event types so it might be that we're not really getting much extra value from extracting them. They'd need to be available to the renderers to make it worthwhile. So all I have is maybe pulling them out to some YAML file.

Let's see where else we use custom events before worrying about that then.

jdbann commented on app/javascript/controllers/map/tokens_controller.js:
```diff
@@ -111,15 +114,32 @@ export default class extends Controller {
+    var draggedOver = event.target
+    var dragging = this.tokenTargets.find(token => token.dataset.beingDragged)
+    if (this.hasDrawerTarget && draggedOver.parentNode == this.drawerTarget && dragging) {
+      this.drawerTarget.insertBefore(dragging, draggedOver)
```
Ooh I like this! What a handy way to handle rearranging!

jdbann commented on app/javascript/controllers/map/tokens_controller.js:
```diff
@@ -139,6 +159,9 @@ export default class extends Controller {
 
     const token = this.findToken(tokenId)
+    if (this.hasDrawerTarget) {
+      this.drawerTarget.dispatchEvent(new CustomEvent('close'))
```
If I follow this correctly, the token drawer will close every time a token is dropped on the map?

Let's see how that feels in practice as it might be a pain for adding multiple creatures

cpytel replied:
That's right, ultimately I thought it should stay open for a bit and then close after some time of not interacting with it, so I decided to do this for now and then improve later.

I also included in my work on token selection, for which a PR is upcoming soon, a keyboard shortcut to open and close the token drawer.

jdbann replied:
Happy to see how this works out. Keyboard shortcuts will be grand 👍 and the only other way I was thinking was clicking on the map could close any open drawers

jdbann commented on app/views/maps/_token_drawer.html.erb:
```diff
@@ -0,0 +1,7 @@
+<div class="current-map__token-drawer token-drawer flex flex-row items-center justify-start select-none bg-gray-500 mx-4 p-2 rounded-lg border whitespace-no-wrap overflow-x-auto overflow-y-hidden shadow-lg" data-target="map--tokens.drawer toggle.content" data-action="dragover->map--tokens#dragOverDrawer drop->map--tokens#dropOnDrawer open->toggle#open close->toggle#close" data-toggle-target="toke ...
```
Tailwind + Stimulus makes for some very long tags. The advantage of inspecting the DOM to look at the current state is impeded a bit as a result. Not for this PR obviously but I wonder if I should put some time into refactoring the styling

jdbann commented on spec/support/system_test.rb:
```diff
@@ -1,6 +1,11 @@
+      ".map-selector.show",
+      ".current-map__token-drawer",
+      ".current-map__token-drawer.show"
+    ].join(",")
```
Shall we just go `true` for this and disable all animation?

cpytel replied:
I can check. I thought I had originally tried that and it caused other specs to fail, but it's been long enough now that I'm not certain.

jdbann replied:
Just a thought for the future 👍

cpytel replied:
Yes, I just tried it and remember now. Pointing things out on the map is entirely a css animation and so disabling all animations causes those specs to fail. I've tried changing this to `true` and then on those pointer specs doing

```
around do |example|
    Capybara.disable_animation = false
    example.run
    Capybara.disable_animation = true
  end
```

But that doesn't seem to work when the whole suite is run, and I can't find a combination of a global and a spec-specific thing that works when the whole suite is run. Let me know if you have any ideas, but for now I think I'll leave this as is.

jdbann commented:
> @jdbann any issue with me merging this this weekend so it's in place for our next session?

Nope, looks great to me! I've had a busy couple of weeks and last weekend was lost to camping so sorry for the slow turnaround on this one
